### PR TITLE
[TACHYON-1049] Skeleton for testing master concurrency performance

### PR DESCRIPTION
**Pull request by uronce-cc**
https://tachyon.atlassian.net/browse/TACHYON-1049

### CODE REVIEW

File: bin/tachyon [hunk 1]
```diff
@@ -18,6 +18,9 @@ function printUsage {
   echo -e "  runTests      \t Run all end-to-end tests on a Tachyon cluster."
   echo -e "  journalCrashTest\t Test the Master Journal System in a crash scenario. Try 'tachyon journalCrashTest -help' for more help."
   echo -e "                  \t NOTE: This command will stop the existing server and creates a new one!"
+  echo -e "  masterConcurrencyPerf\t Performance tests for concurrent master operations."
+  echo -e "                       \t NOTE: This command will restart a new Tachyon cluster locally with one master and one worker,"
+  echo -e "                       \t and underfs storage directory will be deleted and recreated, journal directory will be reformatted."
```
Comment (calvinjia) [inline bin/tachyon]: Why can't the benchmark run on an existing Tachyon cluster and write the files to a directory like `/masterConcurrencyPerf`?

File: examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+import tachyon.client.file.TachyonFileSystem;
+import tachyon.util.io.PathUtils;
+
+/**
+ * Contains performance test cases where a bunch of Tachyon clients concurrently
+ * operates on Tachyon master.
+ */
+public final class MasterConcurrencyPerformance {
+  private static enum OperationType {
+    CREATE_EMPTY_FILE,
```
Comment (calvinjia) [inline examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java]: Are you planning to add other operations?

File: examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+    /**
+     * New operation types should first be added to {@link OperationType}, and then register the
+     * {@link OperationCallable} for this type below in this static block, just use {@link #tfs}
+     * when {@link TachyonFileSystem} is needed.
+     */
+    sOperationRegistry.put(OperationType.CREATE_EMPTY_FILE, new OperationCallable() {
+      @Override
+      public void call(TachyonURI path) throws Exception {
+        tfs.create(path);
```
Comment (calvinjia) [inline examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java]: Should this be `createEmptyFile`?

File: examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    public ClientThread(OperationType opType, List<TachyonURI> paths) {
+      mOpType = opType;
+      mOp = sOperationRegistry.get(opType);
+      mPaths = paths;
+      mState = State.READY;
+    }
+
+    @Override
+    public void run() {
+      mStartTimeMs = System.currentTimeMillis();
```
Comment (calvinjia) [inline examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java]: Should state be set to running here?

File: examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+   */
+  private static void createEmptyFilesUnderRoot(int nFiles, int nClients)
+      throws InterruptedException {
+    int nFilesEachThread = nFiles / nClients;
+    List<ClientThread> clients = Lists.newArrayListWithExpectedSize(nClients);
+    for (int client = 0; client < nClients; client ++) {
+      int files = client == 0 ? nFiles - nFilesEachThread * (nClients - 1) : nFilesEachThread;
+      List<TachyonURI> paths = Lists.newArrayListWithExpectedSize(nFiles);
+      for (int file = 0; file < files; file ++) {
+        paths.add(new TachyonURI(PathUtils.concatPath("/", client, file)));
```
Comment (calvinjia) [inline examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java]: Won't this hit duplicates for client and file > 9?

**Comment by calvinjia** [comment]
@uronce-cc Thanks for the patch, I left a few comments.

File: examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java (hunk 2 above)
Comment (uronce-cc) [inline examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java]: Yes, like `DELETE_FILE`.

File: examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java (hunk 4 above)
Comment (uronce-cc) [inline examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java]: added.

File: examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java (hunk 3 above)
Comment (uronce-cc) [inline examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java]: I think the API has been changed in master?

File: examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java (hunk 5 above)
Comment (uronce-cc) [inline examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java]: concatPath("/", a, b) => /a/b

File: bin/tachyon (hunk 1 above)
Comment (uronce-cc) [inline bin/tachyon]: Ok, I'll keep the strategy the same as runTests.

File: examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java (hunk 5 above)
Comment (calvinjia) [inline examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java]: oh right, that works then

File: examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java (hunk 3 above)
Comment (calvinjia) [inline examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java]: Oh seems like that utility method was lost. I don't think this create will complete the file, but I guess it doesn't matter too much for the perf benchmark.

File: bin/tachyon (hunk 1 above)
Comment (calvinjia) [inline bin/tachyon]: sounds good, thanks

**Comment by uronce-cc** [comment]
@calvinjia Comments resolved, PTAL.

File: examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+/**
+ * Contains performance test cases where a bunch of Tachyon clients concurrently
+ * operates on Tachyon master.
+ */
+public final class MasterConcurrencyPerformance {
+  private static enum OperationType {
+    CREATE_EMPTY_FILE,
+  }
+
+  private static interface OperationCallable {
```
Comment (yupeng9) [inline examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java]: You can use `Function` in guava
http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/base/Function.html

File: examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+      }
+    });
+  }
+
+  /**
+   * A thread containing Tachyon client to do one kind of operation on a list of paths sequentially,
+   * a bunch of such threads can run concurrently to simulate concurrent access to Tachyon master.
+   */
+  private static class ClientThread extends Thread {
+    private static enum State {
```
Comment (yupeng9) [inline examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java]: Do you plan to have all the concurrent tests in this? If not, then this perhaps should not be `private`?

File: examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+  }
+
+  /**
+   * Default root directory to put all files generated by this performance test, can be changed by
+   * command line option.
+   */
+  private static String sTestRootDir = "/tachyonMasterConcurrencyPerf";
+
+  /**
+   * Deletes the dir if it already exists, then, create the empty dir.
```
Comment (yupeng9) [inline examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java]: create->creates

**Comment by calvinjia** [comment]
LGTM

File: examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java (hunk 7 above)
Comment (uronce-cc) [inline examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java]: Yeah, I plan to have all concurrent tests in this.

File: examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java (hunk 8 above)
Comment (uronce-cc) [inline examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java]: done.

File: examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java (hunk 6 above)
Comment (uronce-cc) [inline examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java]: Thanks for hint.

**Comment by uronce-cc** [comment]
@yupeng9 Thanks for the comments! PTAL.

File: examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java (hunk 7 above)
Comment (yupeng9) [inline examples/src/main/java/tachyon/examples/MasterConcurrencyPerformance.java]: SGTM

**Comment by yupeng9** [comment]
Jenkins, test this please

**Comment by yupeng9** [comment]
LGTM. Thanks for adding the starting point of the concurrency test.

**Comment by uronce-cc** [comment]
Jenkins, test this please

**Comment by uronce-cc** [comment]
@calvinjia @yupeng9 PTAL

**Comment by yupeng9** [comment]
LGTM

**Comment by uronce-cc** [comment]
Will port to a standalone repo instead of adding to examples.

**Pull request closed without merging**